data_generator.py

```diff
@@ -0,0 +1,76 @@
+# Holy resources:
+# https://github.com/virafpatrawala/COCO-Semantic-Segmentation/blob/master/COCOdataset_SemanticSegmentation_Demo.ipynb
+# https://mahmoudyusof.github.io/facial-keypoint-detection/data-generator/
+# https://blog.paperspace.com/data-augmentation-for-object-detection-building-input-pipelines/
+
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import cv2
+import matplotlib.image as mpimg
+import os
+import tensorflow as tf
+import skimage.io as io
+from tensorflow.keras.utils import Sequence
+
+class DataGenerator(Sequence):
+
+  def __init__(self, csv_file, base_dir, output_size, shuffle=False, batch_size=10):
```

Comment:
need to pass in `num_hg_blocks` since our labels need to be duplicated based on the number of hourglass blocks (and thus the number of intermediate supervision loss calculations we need to do)

Reply:
@robertklee 
So to clarify, the desired output shape of the data generator is:

train_gen[0] --> X_batch, y_batch --> (batch_size,256,256,3), (batch_size,num_hg_blocks,3*17)

Reply:
Is this duplication of data necessary for the model? Any reason why it can't reuse labels?

Reply:
> Is this duplication of data necessary for the model? Any reason why it can't reuse labels?

necessary because keras doesn't know they're the same labelling. I think it's just easier for keras to handle if we duplicate the ground truth labelling. they probably also do some fancy multi-core stuff with it.

Reply:
> train_gen[0] --> X_batch, y_batch --> (batch_size,256,256,3), (batch_size,**num_hg_blocks**, **num_hg_blocks** *17)
See inline changes

Reply:
X is right, y should be:

(self.batch_size, constants.output_dim, self.num_hg_blocks * constants.num_keypoints)

Output dim is the 64x64 heat map our model outputs (stored in constants.py)
We generate a heat map for *each* joint. so for a single ground truth evaluation, we have 64x64x17.

Since we do intermediate supervision, we duplicate that n times, where n=num hg blocks. So this turns each image label to 64x64x(17*num_hg_blocks).

Then we get the final size of batch_size x *output_dimension x (17*num_hg_blocks)

Reply:
This will be used by keras for MSE evaluation.

Reply:
Thanks for the clarification. Data duplication for keras makes sense then. Looks like the pipeline won't be usable till we get a heatmap label generator then.

Reply:
I can tackle that tomorrow

Reply:
Heatmap gen has been added, output shape adjusted to:
`y.shape = (self.batch_size, constants.output_dim, self.num_hg_blocks * constants.num_keypoints)`